Client/Frontend/Browser/TabManager.swift

```diff
@@ -159,6 +168,45 @@ class TabManager : NSObject {
 
         return nil
     }
+    
+    weak var appStateDelegate: AppStateDelegate?
+    
+    private func updateAppState() {
+        let state = mainStore.updateState(.TabTray(tabTrayState: TabTrayState(isPrivate: self.isInPrivateMode)))
+        self.appStateDelegate?.appDidUpdateState(state)
+    }
+    
+    func authorisePrivateMode(navigationController: UINavigationController, toRemainInPrivateMode reverseAuthorisationRequirement: Bool = false) -> Success {
+        if self.isInPrivateMode != reverseAuthorisationRequirement {
```

Comment:
If this condition was merged into the guard statement, it would have to come after the `KeychainWrapper.authenticationInfo()` call, which would lead to an extra unnecessary set of calls, wouldn't it?